pallets/token-mux/src/lib.rs

```diff
@@ -0,0 +1,377 @@
+// Copyright 2023 Centrifuge Foundation (centrifuge.io).
+//
+// This file is part of the Centrifuge chain project.
+// Centrifuge is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version (see http://www.gnu.org/licenses).
+// Centrifuge is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+// TODO(william): Add pallet description
+
+#[cfg(test)]
+mod mock;
+
+#[cfg(test)]
+pub(crate) mod tests;
+
+#[cfg(feature = "runtime-benchmarks")]
+mod benchmarking;
+
+pub mod weights;
+
+pub use cfg_traits::TokenSwaps;
+pub use pallet::*;
+pub use weights::WeightInfo;
+
+#[frame_support::pallet]
+pub mod pallet {
+
+	use cfg_traits::{OrderRatio, TokenSwaps};
+	use cfg_types::{orders::OrderInfo, tokens::CustomMetadata};
+	use frame_support::{
+		pallet_prelude::{DispatchResult, *},
+		traits::{
+			fungibles,
+			fungibles::Mutate,
+			tokens::{Fortitude, Precision, Preservation},
+		},
+		PalletId,
+	};
+	use frame_system::pallet_prelude::{OriginFor, *};
+	use orml_traits::asset_registry::{self, Inspect as _};
+	use sp_arithmetic::{
+		traits::{AtLeast32BitUnsigned, EnsureDiv, EnsureMul},
+		FixedPointOperand,
+	};
+	use sp_runtime::traits::{AccountIdConversion, EnsureFixedPointNumber, One};
+
+	use super::*;
+
+	pub type BalanceOf<T> = <<T as Config>::Tokens as fungibles::Inspect<
+		<T as frame_system::Config>::AccountId,
+	>>::Balance;
+
+	const STORAGE_VERSION: StorageVersion = StorageVersion::new(0);
+
+	#[pallet::pallet]
+	#[pallet::storage_version(STORAGE_VERSION)]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+		#[pallet::constant]
+		type PalletId: Get<PalletId>;
+
+		/// The source of truth for the existence and potential local
+		/// representation of assets.
+		type AssetRegistry: asset_registry::Inspect<
+			CustomMetadata = CustomMetadata,
+			AssetId = Self::CurrencyId,
+		>;
+
+		/// Balance type for incoming values
+		type BalanceIn: Member
+			+ Parameter
+			+ FixedPointOperand
+			+ AtLeast32BitUnsigned
+			+ EnsureMul
+			+ EnsureDiv
```

Comment:
NIT. IIRC, the `Ensure*` bounds already are in `AlLeast32BitUnsigned` (several ocurrences)